runtime/src/main/cpp/Memory.cpp

```diff
@@ -2056,51 +2056,81 @@ OBJ_GETTER(initSharedInstance,
 #endif  // KONAN_NO_THREADS
 }
 
+/**
+ * We keep thread affinity and reference value based cookie in the atomic references, so that
+ * repeating read operation of the same value do not lead to the repeating rememberNewContainer() operation.
+ * We must invalidate cookie after the local GC, as otherwise fact that container of the `value` is retained
+ * may change, if the last reference to the value read is lost during GC and we re-read same value from
+ * the same atomic reference. Thus we also include last GC timestamp into the cookie.
+ */
+inline int32_t computeCookie(ObjHeader* value) {
+  auto* state = memoryState;
+  auto timestamp = state->lastGcTimestamp;
```

Comment:
Timestamp isn't reliable much as a unique-per-mutation-cycle value. Why not use a dedicated counter for this?

Reply:
Given its microsecond resolution it shall work as such, but indeed could make it a counter.